File: ChessLab/Laboratory/LabStats.swift
import Foundation

/// Résultat d'une partie du point de vue du « camp A » — la force
/// configurée à gauche dans les réglages du Laboratoire. L'alternance des
/// couleurs (A joue tantôt Blanc, tantôt Noir) est déjà résolue en amont
/// par ``LabViewModel`` ; ici A désigne toujours le même moteur/réglage,
/// indépendamment de la couleur, ce qui est exactement ce qu'il faut pour
/// estimer un écart Elo non biaisé par l'avantage des Blancs.
enum LabGameResult: String, Codable, Equatable {
    case winA
    case draw
    case winB
}

/// Statistiques agrégées d'une série Laboratoire (Stockfish vs Stockfish).
///
/// Entièrement pur et testable sans moteur : toutes les formules (score,
/// écart Elo, intervalle de confiance à 95 %, LOS) ne dépendent que des
/// comptes W/N/D et de la longueur des parties.
struct LabStats: Equatable {
    /// Longueurs de partie en demi-coups (plies), une entrée par partie
    /// terminée — sert aussi à l'histogramme.
    let plyCounts: [Int]
    let winsA: Int
    let draws: Int
    let winsB: Int

    init(results: [LabGameResult], plyCounts: [Int]) {
        self.plyCounts = plyCounts
        winsA = results.filter { $0 == .winA }.count
        draws = results.filter { $0 == .draw }.count
        winsB = results.filter { $0 == .winB }.count
    }

    var games: Int { winsA + draws + winsB }

    /// Score de A au sens échiquéen : 1 par gain, ½ par nulle.
    var score: Double {
        guard games > 0 else { return 0 }
        return (Double(winsA) + 0.5 * Double(draws)) / Double(games)
    }

    var scorePercent: Double { score * 100 }

    var averagePlies: Double {
        guard !plyCounts.isEmpty else { return 0 }
        return Double(plyCounts.reduce(0, +)) / Double(plyCounts.count)
    }

    /// Longueur moyenne en coups complets (un coup = deux demi-coups).
    var averageMoves: Double { averagePlies / 2 }

    /// Écart Elo estimé de A par rapport à B : −400·log₁₀(1/score − 1).
    /// `nil` quand le score vaut 0 ou 1 (écart théoriquement infini — trop
    /// peu de données pour trancher).
    var eloDifference: Double? { Self.elo(fromScore: score) }

    /// Intervalle de confiance à 95 % sur l'écart Elo, dérivé de l'erreur
    /// standard du score. `nil` s'il n'est pas calculable (< 2 parties, ou
    /// bornes de score dégénérées).
    var elo95ConfidenceInterval: (low: Double, high: Double)? {
        guard let se = scoreStandardError else { return nil }
        let loScore = clampScore(score - 1.96 * se)
        let hiScore = clampScore(score + 1.96 * se)
        guard let low = Self.elo(fromScore: loScore), let high = Self.elo(fromScore: hiScore) else { return nil }
        return (low, high)
    }

    /// Likelihood of Superiority : probabilité que A soit réellement plus
    /// fort que B, estimée sur les seules parties décisives (modèle
    /// binomial normal-approximé). 0,5 quand il n'y a aucune décision.
    var likelihoodOfSuperiority: Double {
        let decisive = winsA + winsB
        guard decisive > 0 else { return 0.5 }
        let x = Double(winsA - winsB) / (sqrt(2.0) * sqrt(Double(decisive)))
        return 0.5 * (1 + erf(x))
    }

    // MARK: Interne

    /// Erreur standard du score moyen (résultats codés 1 / ½ / 0).
    private var scoreStandardError: Double? {
        guard games > 1 else { return nil }
        let sumSquares = Double(winsA) * 1 + Double(draws) * 0.25 // + winsB * 0
        let mean = score
        let variance = max(0, sumSquares / Double(games) - mean * mean)
        return sqrt(variance / Double(games))
    }

    private func clampScore(_ s: Double) -> Double {
        min(max(s, 1e-6), 1 - 1e-6)
    }

    static func elo(fromScore s: Double) -> Double? {
        guard s > 0, s < 1 else { return nil }
        return -400 * log10(1 / s - 1)
    }
}

/// Un point de la courbe de progression du Laboratoire : score cumulé de A
/// (en %) après la partie `game`, avec sa bande de confiance à 95 %.
struct LabProgressPoint: Identifiable, Equatable {
    let game: Int          // numéro de partie, 1-based
    let scorePercent: Double
    let ciLow: Double       // borne basse du score % (95 %), dans [0 ; 100]
    let ciHigh: Double
    let result: LabGameResult
    var id: Int { game }
}

extension LabStats {
    /// Courbe de progression : un point par partie jouée, score cumulé de A
    /// (%) + intervalle de confiance à 95 % qui se resserre à mesure que la
    /// série devient significative. Pur et testable.
    static func progression(of games: [LabCompletedGame]) -> [LabProgressPoint] {
        var points: [LabProgressPoint] = []
        points.reserveCapacity(games.count)
        for index in games.indices {
            let prefix = games[0...index]
            let stats = LabStats(results: prefix.map(\.labResult), plyCounts: prefix.map(\.plyCount))
            let margin = 1.96 * (stats.scoreStandardError ?? 0)
            points.append(
                LabProgressPoint(
                    game: index + 1,
                    scorePercent: stats.score * 100,
                    ciLow: max(0, stats.score - margin) * 100,
                    ciHigh: min(1, stats.score + margin) * 100,
                    result: games[index].labResult
                )
            )
        }
        return points
    }
}

File: android/app/src/main/java/com/chesslab/lab/LabStats.kt
package com.chesslab.lab

import kotlin.math.log10
import kotlin.math.max
import kotlin.math.min
import kotlin.math.sqrt

/**
 * Le résultat d'une partie DU POINT DE VUE DU CAMP A — celui réglé à gauche.
 *
 * L'alternance des couleurs (A joue tantôt Blanc, tantôt Noir) est déjà
 * résolue en amont : ici A désigne toujours le même réglage, quelle que soit
 * sa couleur. C'est exactement ce qu'il faut pour estimer un écart Elo que
 * l'avantage du trait ne fausse pas.
 */
enum class LabGameResult { winA, draw, winB }

/** Une partie terminée de la série, telle qu'on la garde pour le bilan. */
data class LabCompletedGame(
    val index: Int,
    /** A jouait-il les Blancs dans CETTE partie ? */
    val aWasWhite: Boolean,
    /** Le résultat côté échiquier : « 1-0 », « 0-1 », « 1/2-1/2 ». */
    val pgnResult: String,
    /** La raison de la fin, en clair. */
    val reasonLabel: String,
    val plyCount: Int,
    val pgn: String,
) {
    /** Le résultat rapporté au camp A, indépendant de la couleur. */
    val labResult: LabGameResult
        get() = when (pgnResult) {
            "1-0" -> if (aWasWhite) LabGameResult.winA else LabGameResult.winB
            "0-1" -> if (aWasWhite) LabGameResult.winB else LabGameResult.winA
            else -> LabGameResult.draw
        }
}

/**
 * Le bilan agrégé d'une série. Pendant de `LabStats.swift`.
 *
 * Entièrement PUR : score, écart Elo, intervalle de confiance et LOS ne
 * dépendent que des comptes et de la longueur des parties. C'est ce qui rend
 * ces chiffres vérifiables sans lancer un seul moteur — et ils méritent de
 * l'être, puisque c'est sur eux qu'on décide si une série a tranché.
 */
data class LabStats(
    /** Longueurs en DEMI-COUPS, une entrée par partie terminée. */
    val plyCounts: List<Int>,
    val winsA: Int,
    val draws: Int,
    val winsB: Int,
) {
    val games: Int get() = winsA + draws + winsB

    /** Le score de A au sens échiquéen : 1 par gain, ½ par nulle. */
    val score: Double
        get() = if (games == 0) 0.0 else (winsA + 0.5 * draws) / games

    val scorePercent: Double get() = score * 100

    val averagePlies: Double
        get() = if (plyCounts.isEmpty()) 0.0 else plyCounts.sum().toDouble() / plyCounts.size

    /** La longueur moyenne en COUPS COMPLETS (un coup = deux demi-coups). */
    val averageMoves: Double get() = averagePlies / 2

    /**
     * L'écart Elo estimé de A par rapport à B : −400·log₁₀(1/score − 1).
     *
     * `null` quand le score vaut 0 ou 1 : l'écart y est théoriquement infini,
     * ce qui veut simplement dire qu'il y a trop peu de données pour trancher.
     */
    val eloDifference: Double? get() = elo(score)

    /**
     * L'intervalle de confiance à 95 % sur l'écart Elo. `null` s'il n'est pas
     * calculable — moins de deux parties, ou bornes dégénérées.
     */
    val elo95ConfidenceInterval: Pair<Double, Double>?
        get() {
            val se = scoreStandardError ?: return null
            val low = elo(clampScore(score - 1.96 * se)) ?: return null
            val high = elo(clampScore(score + 1.96 * se)) ?: return null
            return low to high
        }

    /**
     * LOS — la probabilité que A soit RÉELLEMENT plus fort que B, et non que
     * l'écart observé vienne du hasard.
     *
     * Estimée sur les seules parties DÉCISIVES : une nulle ne dit rien sur qui
     * est le plus fort. 0,5 quand rien n'a été décidé.
     */
    val likelihoodOfSuperiority: Double
        get() {
            val decisive = winsA + winsB
            if (decisive == 0) return 0.5
            val x = (winsA - winsB) / (sqrt(2.0) * sqrt(decisive.toDouble()))
            return 0.5 * (1 + erf(x))
        }

    /**
     * L'erreur standard du score moyen (résultats codés 1 / ½ / 0).
     *
     * Deux correctifs venus d'iOS, qui changent les chiffres affichés et
     * méritent donc d'être dits :
     * - **Bessel** (`n − 1`, pas `n`) : ces `n` parties ESTIMENT une force
     *   inconnue ; diviser par `n` sous-estime l'incertitude — 5 % trop
     *   étroit à dix parties, négligeable à cent ;
     * - **terme de continuité (Wilson)** : sans lui, un échantillon dont
     *   toutes les parties partagent le même résultat (deux nulles d'affilée,
     *   par exemple) donne une variance EXACTEMENT nulle, donc un intervalle
     *   de largeur nulle : une fausse certitude à 95 % après trois parties.
     *   `z²/(4n²)` est le terme que Wilson ajoute à cette même place.
     */
    internal val scoreStandardError: Double?
        get() {
            if (games <= 1) return null
            val n = games.toDouble()
            val sumSquares = winsA * 1.0 + draws * 0.25       // + winsB * 0
            val sampleVariance = max(0.0, (sumSquares - n * score * score) / (n - 1))
            val continuityTerm = (1.96 * 1.96) / (4 * n * n)
            return sqrt(sampleVariance / n + continuityTerm)
        }

    private fun clampScore(s: Double): Double = min(max(s, 1e-6), 1 - 1e-6)

    companion object {
        fun of(results: List<LabGameResult>, plyCounts: List<Int>) = LabStats(
            plyCounts = plyCounts,
            winsA = results.count { it == LabGameResult.winA },
            draws = results.count { it == LabGameResult.draw },
            winsB = results.count { it == LabGameResult.winB },
        )

        fun elo(score: Double): Double? =
            if (score <= 0.0 || score >= 1.0) null else -400 * log10(1 / score - 1)

        /**
         * La courbe de progression : un point par partie, score CUMULÉ de A
         * avec sa bande de confiance — qui se resserre à mesure que la série
         * devient significative. Tant qu'elle chevauche les 50 %, rien n'est
         * tranché ; c'est son rétrécissement qui dit quand s'arrêter.
         */
        fun progression(games: List<LabCompletedGame>): List<LabProgressPoint> =
            games.indices.map { index ->
                val prefix = games.subList(0, index + 1)
                val stats = of(prefix.map { it.labResult }, prefix.map { it.plyCount })
                val margin = 1.96 * (stats.scoreStandardError ?: 0.0)
                LabProgressPoint(
                    game = index + 1,
                    scorePercent = stats.score * 100,
                    ciLow = max(0.0, stats.score - margin) * 100,
                    ciHigh = min(1.0, stats.score + margin) * 100,
                    result = games[index].labResult,
                )
            }

        /**
         * La fonction d'erreur, que la bibliothèque standard de Kotlin n'a
         * pas — contrairement au C dont Swift hérite `erf`. Approximation
         * d'Abramowitz & Stegun 7.1.26, exacte à 1,5·10⁻⁷ : bien au-delà de
         * ce qu'un pourcentage affiché demande.
         */
        internal fun erf(x: Double): Double {
            val sign = if (x < 0) -1.0 else 1.0
            val a = kotlin.math.abs(x)
            val t = 1.0 / (1.0 + 0.3275911 * a)
            val y = 1.0 - (((((1.061405429 * t - 1.453152027) * t) + 1.421413741) * t - 0.284496736) * t +
                0.254829592) * t * kotlin.math.exp(-a * a)
            return sign * y
        }
    }
}

/** Un point de la courbe de progression : le score cumulé de A après `game`. */
data class LabProgressPoint(
    /** Le numéro de partie, à partir de 1. */
    val game: Int,
    val scorePercent: Double,
    /** Les bornes du score en %, à 95 %, dans [0 ; 100]. */
    val ciLow: Double,
    val ciHigh: Double,
    val result: LabGameResult,
)
